onnxruntime/core/providers/cuda/math/inverse.cc

```diff
@@ -0,0 +1,153 @@
+        // Convert from MLFloat16(half) to float
+        Impl_Cast<CudaT, float>(reinterpret_cast<const CudaT*>(input.Data<MLFloat16>()), input_workspace.get(), input_count);
+      } else {
+        CUDA_RETURN_IF_ERROR(cudaMemcpy(input_workspace.get(), input.Data<float>(), sizeof(float) * input_count,
+                                        cudaMemcpyDeviceToDevice));
+      }
+      IAllocatorUniquePtr<float*> matrix_ptrs = inst->GetScratchBuffer<float*>(n_batches);
+      ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(input_workspace.get(), num_batches, rows, matrix_ptrs));
+      // Do LU factorization
+      CUBLAS_RETURN_IF_ERROR(cublasSgetrfBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), info.get(), n_batches));
+      ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+
+      // Need to compute ptrs for output buffers
+      // Output for MLFloat
+      IAllocatorUniquePtr<float*> output_ptrs = inst->GetScratchBuffer<float*>(n_batches);
+      if (std::is_same<T, MLFloat16>::value) {
+        IAllocatorUniquePtr<float> ml_float_output = inst->GetScratchBuffer<float>(input_count);
+        ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(ml_float_output.get(), num_batches, rows, output_ptrs));
+        // Do the inverse
+        CUBLAS_RETURN_IF_ERROR(cublasSgetriBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), output_ptrs.get(), dim, info.get(), n_batches));
+        ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+        // Copy the result to output with casting
+        Impl_Cast<float, CudaT>(ml_float_output.get(), reinterpret_cast<CudaT*>(output.MutableData<MLFloat16>()), input_count);
+        // We are done here
+      } else {
+        ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(output.MutableData<float>(), num_batches, rows, output_ptrs));
+        // Do the inverse
+        CUBLAS_RETURN_IF_ERROR(cublasSgetriBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), output_ptrs.get(), dim, info.get(), n_batches));
+        ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+        // We are done here
+      }
+    } else if (std::is_same<T, double>::value) {
+      IAllocatorUniquePtr<double> input_workspace = inst->GetScratchBuffer<double>(static_cast<int>(input_count));
+      CUDA_RETURN_IF_ERROR(cudaMemcpy(input_workspace.get(), input.Data<double>(), sizeof(double) * input_count,
+                                      cudaMemcpyDeviceToDevice));
+
+      IAllocatorUniquePtr<double*> matrix_ptrs = inst->GetScratchBuffer<double*>(n_batches);
+      ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<double>(input_workspace.get(), num_batches, rows, matrix_ptrs));
+      // Do LU factorization
+      CUBLAS_RETURN_IF_ERROR(cublasDgetrfBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), info.get(), n_batches));
+      ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
```

Comment:
There is some external discussion as to whether this approach is performant if there was a single large matrix to be inverted (https://stackoverflow.com/questions/37731103/cublas-matrix-inverse-much-slower-than-matlab). Basically, the approach you are taking is conducive for inverting a batch of smaller matrices. Quoting  cuBlas documentation - "This function is intended to be used for matrices of small sizes where the launch overhead is a significant factor." Maybe there must be a plan to deal with a single large matrix.

Reply:
This means it is optimized for small matrix overhead as well as solving large matrices. This is what my research says.

---
In reply to: [406944729](https://github.com/microsoft/onnxruntime/pull/3485#discussion_r406944729) [](ancestors = 406944729)